### PR TITLE
Add support for linking CUDA C / C++ with `@cuda.jit` kernels

### PR DESCRIPTION
**Pull request by gmarkall**
This adds support for linking CUDA C/C++ sources using NVRTC. One links them in the same way as PTX files, i.e.:

```python
@cuda.jit(link=['source.cu'])
def f(x):
    # bar() is defined in source.cu
    bar(x)
```

Some design decisions I've made for this initial implementation:

- This uses the NVIDIA binding to access NVRTC, so that issues around locating NVRTC are delegated to it, rather than adding logic to Numba that duplicates this functionality (and is potentially difficult and error-prone to implement and test).
- There is no support for the ctypes API. Ideally I'd avoid adding it, with the current plan being to eventually default to the NVIDIA bindings and remove the ctypes bindings in the long run.
- Whilst not all CUDA sources require the CUDA include directory, some do. Towards providing this, Numba assumes a default location for the include path, and allows it to be set if necessary though `NUMBA_CUDA_INCLUDE_PATH`. Some more sophisticated logic could be added to detect the include path in future if this seems necessary / useful to do and a standard / sensible way to do it can be found. (An example of requiring the CUDA includes is in numba/numba-examples#40, where cuRAND includes are used).
- No support for adding additional include paths is present for this initial iteration - sources must only rely on CUDA includes for now. Support for specifying include paths may be added in future.

Based on #7619, which fixes issues with linking when lazily compiling (though it is not strictly necessary for this PR, it makes sense to consider on top of it).

### CODE REVIEW

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by seibert** [comment]
Will this provide a solution for accessing cuRandom?  Would love to (slowly) deprecate our existing CUDA random number generator.

**Comment by gmarkall** [comment]
@seibert It makes it easier to call cuRAND (it was always possible, but you'd have to compile to PTX first). Here's an example: https://github.com/numba/numba-examples/pull/40 - I think we could work through the cuRAND API and provide a nice wrapper for it using this functionality, it's likely just a case of finding the hours to put into it.

(is cuRandom something different? My searches for it turn up mostly `pycuda.curandom`)

To compile code using cuRAND, I also had to make a small hack on this PR, to include the CUDA include directory, e.g.:

```diff
diff --git a/numba/cuda/cudadrv/driver.py b/numba/cuda/cudadrv/driver.py
index 81e7d608a..a78bcc4f1 100644
--- a/numba/cuda/cudadrv/driver.py
+++ b/numba/cuda/cudadrv/driver.py
@@ -2810,9 +2810,10 @@ class CudaPythonLinker(Linker):
             major, minor = dev.compute_capability
 
         arch = bytes(f'--gpu-architecture=compute_{major}{minor}', 'ascii')
+        include = b'-I/usr/local/cuda-11.4/include'
 
         # rdc is needed to prevent device functions being optimized away
-        opts = [arch, b'-rdc', b'true']
+        opts = [arch, b'-rdc', b'true', include]
         err, = nvrtc.nvrtcCompileProgram(prog, len(opts), opts)
         compile_error = (err == nvrtc.nvrtcResult.NVRTC_ERROR_COMPILATION)
 
```

so I probably need to handle finding an appropriate CUDA include dir in this PR too.

**Comment by seibert** [comment]
oops, I got the name wrong.  Yes, I meant cuRAND.  :)

**Comment by gmarkall** [comment]
This will also help enable https://github.com/numba/numba/issues/7528

**Comment by gmarkall** [comment]
Note to anyone who's pulled locally or recently observed changes: I did a force push to squash out some commits where I had a file called `aux.cu`, which will cause pain for Windows users if they ever try to checkout those commits. The history diverges just after commit 7fe9422 (Add NvrtcError to simulator).

**Comment by gmarkall** [comment]
(gpuci fail seems to be due to bad Conda downloads)

gpuci run tests

**Comment by gmarkall** [comment]
Although gpuCI seems to be having a bad day, I'm going to mark this as ready for review because it's certainly in a state to receive feedback. When I get back to work on Monday I'll rerun the build (and hopefully all will be well since the last gpuCI pass I only added documentation and Windows-specific path handling).

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
gpuci run tests

File: docs/source/cuda/cuda_ffi.rst
Comment (stuartarchibald) [inline docs/source/cuda/cuda_ffi.rst:94]: Does something need to be said about:
* The required presence of `nvrtc` on the path (and where to get it).
* How specified files are resolved via paths, am guessing working directory only or full path required?

File: docs/source/cuda/cuda_ffi.rst
Comment (stuartarchibald) [inline docs/source/cuda/cuda_ffi.rst:117]:
```suggestion
- The CUDA include directory will be made available to NVRTC on the include path;
```

File: numba/core/config.py
Comment (stuartarchibald) [inline numba/core/config.py:423]: Should this be `os.path.join(['usr', 'local', 'cuda', 'include'])`?

File: numba/cuda/cudadrv/driver.py
Comment (stuartarchibald) [inline numba/cuda/cudadrv/driver.py:2613]: Is the `name` here e.g. the file name from which the `cu` string was obtained?

File: numba/cuda/cudadrv/driver.py
Comment (stuartarchibald) [inline numba/cuda/cudadrv/driver.py:2782]:
```suggestion
            msg = (f'NVRTC Compilation failure whilst compiling {name}:\n\n'
```

File: numba/cuda/cudadrv/driver.py
Comment (stuartarchibald) [inline numba/cuda/cudadrv/driver.py:2788]:
```suggestion
            msg = f"NVRTC log messages whilst compiling {name}:\n\n{log.decode()}"
```

File: numba/cuda/cudadrv/driver.py
Comment (stuartarchibald) [inline numba/cuda/cudadrv/driver.py:2789]: Should this be a Numba CUDA specific warning class?

File: numba/cuda/decorators.py
Comment (stuartarchibald) [inline numba/cuda/decorators.py:152]: Should this also accept a `numba.core.typing.templates.Signature` such as that derived from composing a signature with `numba.types` types?

File: numba/cuda/tests/cudadrv/data/cuda_include.cu
Comment (stuartarchibald) [inline numba/cuda/tests/cudadrv/data/cuda_include.cu:1]: I think these `.cu` files will need adding to `package_data` in `setup.py`.

File: numba/cuda/tests/doc_examples/ffi/functions.cu
Comment (stuartarchibald) [inline numba/cuda/tests/doc_examples/ffi/functions.cu:1]: I think this `.cu` file will need adding to `package_data` in `setup.py`.

File: numba/cuda/tests/cudadrv/test_linker.py
Comment (stuartarchibald) [inline numba/cuda/tests/cudadrv/test_linker.py:104]: Is `jitlink.cu` missing from this patch?

File: numba/cuda/tests/cudadrv/test_linker.py
Comment (stuartarchibald) [inline numba/cuda/tests/cudadrv/test_linker.py:129]: context managed block should probably have a call to this to filter known internal warnings:
https://github.com/numba/numba/blob/59fa2e335be68148b3bd72a29de3ff011430038d/numba/tests/support.py#L141-L144

File: numba/cuda/tests/doc_examples/test_ffi.py
Comment (stuartarchibald) [inline numba/cuda/tests/doc_examples/test_ffi.py:22]:
```suggestion
        # (here assumed to be in a sub-directory called "ffi")
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for implementing this functionality @gmarkall. On inspection, the code in general looks good, I've left a few minor comments/queries inline. Will manually tests this next to try it out! Thanks again!

File: numba/core/config.py
Comment (gmarkall) [reply numba/core/config.py:423]: Yes, that would be better.

File: numba/cuda/cudadrv/driver.py
Comment (gmarkall) [reply numba/cuda/cudadrv/driver.py:2613]: It could be, but it's not required.

File: numba/cuda/cudadrv/driver.py
Comment (gmarkall) [reply numba/cuda/cudadrv/driver.py:2789]: None of the other warnings in driver.py use a specific class. I think there's not too great a need for this - I'd imagine you always want to resolve the warnings rather than filter them, but I'm open to a suggestion as to what I might be missing / misjudging here.

File: numba/cuda/decorators.py
Comment (gmarkall) [reply numba/cuda/decorators.py:152]: It will already accept one, I think I just got carried away with the docs here. I'll just remove this type documentation since a Numba signature can be a `Signature` or a string.

File: numba/cuda/tests/cudadrv/test_linker.py
Comment (gmarkall) [reply numba/cuda/tests/cudadrv/test_linker.py:104]: It already exists: https://github.com/numba/numba/blob/main/numba/cuda/tests/cudadrv/data/jitlink.cu

File: numba/cuda/tests/cudadrv/test_linker.py
Comment (gmarkall) [reply numba/cuda/tests/cudadrv/test_linker.py:132]: I've now added the call.

File: numba/cuda/tests/doc_examples/ffi/functions.cu
Comment (gmarkall) [reply numba/cuda/tests/doc_examples/ffi/functions.cu:1]: Added.

File: numba/cuda/tests/cudadrv/data/cuda_include.cu
Comment (gmarkall) [reply numba/cuda/tests/cudadrv/data/cuda_include.cu:1]: Added.

File: docs/source/cuda/cuda_ffi.rst
Comment (gmarkall) [reply docs/source/cuda/cuda_ffi.rst]: I've now added a note about the relative / absolute path specification for linked files, and also about the need for NVRTC. I didn't get too specific about NVRTC because I felt it would get needlessly complex and confusing trying to be specific about the possibilities across platforms and versions.

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
@stuartarchibald Many thanks for the review - I have addressed / responded to all comments, so I think this is ready for another look.

File: docs/source/cuda/cuda_ffi.rst
Comment (stuartarchibald) [reply docs/source/cuda/cuda_ffi.rst]: Thanks for adding the path specification. RE the `nvrtc` binary location, this seems reasonable, it's quite an advanced use case and it needing to be on the path would likely be an implicit requirement, how this is resolved appropriately is understandably user specific.

File: numba/cuda/cudadrv/driver.py
Comment (stuartarchibald) [reply numba/cuda/cudadrv/driver.py:2790]: ok, thanks, I think for consistency with the current code base it's fine to leave this as is in this PR, but might be something to consider at a later time.

File: numba/cuda/decorators.py
Comment (stuartarchibald) [reply numba/cuda/decorators.py]: Thanks, seems like that's the easiest fix!

File: numba/cuda/tests/cudadrv/test_linker.py
Comment (stuartarchibald) [reply numba/cuda/tests/cudadrv/test_linker.py:105]: ah yes, thanks.

File: numba/cuda/cudadrv/driver.py
Comment (stuartarchibald) [reply numba/cuda/cudadrv/driver.py:2613]: Should the docstring explain what it is? If it's not required should there be a default?

Comment (gmarkall) [reply numba/cuda/cudadrv/driver.py:2613]: Sorry, I mean it's not required to be the filename from which the cu string was obtained. However, it should be provided and should end in .cu - I've updated the docstring.

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
@stuartarchibald All should be ready for another look now.

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
gpuci run tests

File: numba/cuda/cudadrv/driver.py
Comment (gmarkall) [reply numba/cuda/cudadrv/driver.py:2613]: Since I'm now using splitext (as you suggested OOB) it doesn't really matter what the extension is, so I've removed that requirement from the docstring.

**Comment by gmarkall** [comment]
@stuartarchibald I think this is ready again for re-reviewing the responses to the last couple of comments.

**Comment by stuartarchibald** [comment]
Notes from manual testing:
1. Checking mis-spelling of file name and extension behave reasonably but error handling could be improved.
2. Checking mis-matched function prototypes breaks in an expected way.
3. Checking two `.cu` files defining the same named function is spotted as a multiple definition.
4. CUDA does unsafe conversion of e.g. float64 to float32 if the `.cu` file definition takes a `float32`.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates @gmarkall, couple of minor other things else looks good. Thanks again!

File: numba/cuda/cudadrv/driver.py
Comment (gmarkall) [reply numba/cuda/cudadrv/driver.py:2622]: I think it might as well just try, rather than reimplementing the error-handling logic of `open()`. For example, if you try to link a nonexistent file, the last line of the traceback is pretty clear:

```
FileNotFoundError: [Errno 2] No such file or directory: '/home/gmarkall/numbadev/numba/numba/cuda/tests/doc_examples/ffi/functioddns.cu'
```

Comment (stuartarchibald) [reply numba/cuda/cudadrv/driver.py:2622]: ok, let's leave it as-is. It does state the issue, which is probably sufficient to start a bug report for e.g. the maintainers of a library that is using this functionality.

**Comment by gmarkall** [comment]
gpuci run tests

File: numba/cuda/cudadrv/driver.py
Comment (gmarkall) [reply numba/cuda/cudadrv/driver.py]: Using it now.

File: numba/cuda/cudadrv/driver.py
Comment (gmarkall) [reply numba/cuda/cudadrv/driver.py]: I've changed it anyway, it now raises a `RuntimeError` describing the problem (depending on whether the extension is unknown, or non-existent) and there's a couple of tests for this.

Comment (stuartarchibald) [reply numba/cuda/cudadrv/driver.py]: Thanks!

**Review by stuartarchibald — DISMISSED** [review]
Thanks for the patch and fixes!

**Comment by stuartarchibald** [comment]
> @stuartarchibald I've made updates responding to all the feedback - assuming gpuCI and the buildfarm pass, I'd like to give this another manual check on Windows too, because I've changed the code since I last checked on Windows with the CUDA Python bindings.

@gmarkall No problem, thanks for working through all of the review comments. I'll mark this for a build farm run now as GPU CI has passed, will also wait for your feedback on how a manual test on Windows with the CUDA Python bindings goes. Thanks again!

File: numba/cuda/cudadrv/driver.py
Comment (gmarkall) [reply numba/cuda/cudadrv/driver.py:2730]: I've added a test for this now. The linking cu tests that actually try to link a cu file are skipped if the ctypes binding is in use.

**Comment by gmarkall** [comment]
gpuci run tests

File: numba/cuda/tests/cudadrv/test_linker.py
Comment (stuartarchibald) [inline numba/cuda/tests/cudadrv/test_linker.py:164]: Perhaps:
```suggestion
        msg = ('Linking CUDA source files is not supported with the ctypes '
               'binding')
```
just to make absolutely sure it's that specific exception from linking and not a different but similar one?

**Review by stuartarchibald — APPROVED** [review]
Thanks for adding the additional test.

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cuda_yaml_125`.

**Comment by stuartarchibald** [comment]
> Buildfarm ID: `numba_smoketest_cuda_yaml_125`.

Passed.

**Pull request merged**
